hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/metadata/HazelcastRelMdPrunability.java

```diff
@@ -0,0 +1,174 @@
+                .map(TableField::getName)
+                .collect(Collectors.toSet());
+
+        if (targetTable.partitioningAttributes().isEmpty()) {
+            partitioningColumns.add(QueryPath.KEY);
+        }
+
+        final RexNode filter = hazelcastTable.getFilter();
+        if (!(filter instanceof RexCall)) {
+            return emptyMap();
+        }
+
+        final RexCall call = (RexCall) filter;
+        final var conditionExtractor = new PartitionStrategyConditionExtractor();
+        return conditionExtractor.extractCondition(targetTable, call, partitioningColumns);
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(
+            IndexScanMapPhysicalRel scan,
+            RelMetadataQuery mq
+    ) {
+        // TODO: Implement
+        return emptyMap();
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(Calc calc, RelMetadataQuery mq) {
+        HazelcastRelMetadataQuery query = HazelcastRelMetadataQuery.reuseOrCreate(mq);
+        return query.extractPrunability(calc.getInput());
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(Aggregate agg, RelMetadataQuery mq) {
+        // TODO: Implement
+        return emptyMap();
+    }
+
+    // It is done to support usage of this metadata query during opt phase.
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(RelSubset subset, RelMetadataQuery mq) {
+        HazelcastRelMetadataQuery query = HazelcastRelMetadataQuery.reuseOrCreate(mq);
+        RelNode rel = Util.first(subset.getBest(), subset.getOriginal());
+        return query.extractPrunability(rel);
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(Join rel, RelMetadataQuery mq) {
+        // For any bi-rel (Joins) we (temporarily) are not propagating prunability.
+        return Collections.emptyMap();
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(RelNode rel, RelMetadataQuery mq) {
+        // For any non-mentioned rels, we assume they are prunable and forwards prunability.
+        HazelcastRelMetadataQuery query = HazelcastRelMetadataQuery.reuseOrCreate(mq);
+        Map<String, List<Map<String, RexNode>>> prunability = new HashMap<>();
+        for (int i = 0; i < rel.getInputs().size(); i++) {
+            RelNode input = rel.getInput(i);
+            var extractedPrunability = query.extractPrunability(input);
+            for (final String tableName : extractedPrunability.keySet()) {
+                var tableVariants = extractedPrunability.get(tableName);
+                prunability.putIfAbsent(tableName, new ArrayList<>());
+                prunability.get(tableName).addAll(tableVariants);
```

Comment:
same as [above/below](https://github.com/hazelcast/hazelcast/pull/24813/files#r1263756501):
what if the same table is used in given query multiple times with different predicates and aliases? Table name is not unique on query level

maybe we should track prunablility per RelNode (`RelNode.getId()`)?

Reply:
ok, we collect list of all partition keys from all references to an IMap in given query. This should be enough for member pruning.